Adds AmpAdTemplate

glevitzky opened the pull request:
Adds feature that allows templatized ads to be fetched and rendered via Fast Fetch. 

Sample implementation:

On the page:
```html
<amp-ad type="test" width=320 height=50
  data-request-var-param1="val1"
  data-request-var-param2="val2"
  . . .
  data-request-var-paramn="valn"
  template>
</amp-ad>
````

Note that all attributes of the form `data-request-var-<param_name>=<value>` will be added to the request URL by appending`&<param_name>=<value>` to the registered URL (see below).

In amp-a4a/0.1/template-config.js:
```js
export const NetworkRegistry = {
  // . . .
  myAdNetwork: {
    requestUrl: 'https://www.generic-ad-network.com?optionalStaticParam=foo',
  },
};
```

Some fields will automatically be substituted into the URL. Currently these are:
- height
- width

E.g., something like `...&sz=widthxheight` will be expanded to `...&sz=320x50`, if the size of the slot at request time is 320x50.

Addresses #15464.

jridgewell commented on extensions/amp-a4a/0.1/template-renderer.js:
```diff
@@ -22,27 +22,29 @@ import {getAmpAdTemplateHelper} from './template-validator';
+                .querySelector('template');
+            templateElement.parentNode
+                .replaceChild(renderedElement, templateElement);
+          });
```
👍, my nits are fixed. Thanks!

lannka commented:
High level:

- I suggest to not have the `template` attribute. For real ad network integration, we should let the ad server decide what rendering method it wants.
- How about `type="custom-template"` instead of `type="test"`
- Nit: rename the attributes to `data-request-param-abc-dce`. Also convert the dash separated var name to camel case, as a convention in amp-analytis `data-vars-*`. (strings.js#dashToCamelCase)

glevitzky commented:
@lannka Thanks for the feedback.

* If we do not use `template` attribute, how do we divert in amp-ad to use the correct constructor?
* Easy enough.
* Renamed to `data-request-param...`. Conversion to camel case happens automatically when reading element.dataset, but I've cleaned up the code around this a bit. Not sure how we should handle a case where pub wants a dash-separated param on request, but we can deal with that down the road.

lannka commented on extensions/amp-a4a/0.1/template-renderer.js:
```diff
@@ -22,27 +22,29 @@ import {getAmpAdTemplateHelper} from './template-validator';
  */
 export class TemplateRenderer extends FriendlyFrameRenderer {
```
in general, inheritance like this is discouraged. Let's extract shared logic to a util method instead, and have `TemplateRenderer extends Renderer `. See [here](https://en.wikipedia.org/wiki/Composition_over_inheritance) for more background around why.

glevitzky replied:
Instead of extending FriendlyFrameRenderer, can I keep an instance of it in TemplateRenderer? Since TemplateRenderer eventually renders into a friendly frame, it's unavoidable that I'll need the entirety of FriendlyFrameRenderer's logic.

lannka replied:
totally. i suggest convert FriendlyFrameRenderer into a util class without extending Renderer.

glevitzky replied:
For now I'd prefer to keep it as its own Renderer subclass, so its consistent and uniform with the other renderers. This will especially be useful once I finally get around to tackling the amp-a4a refactor.

lannka replied:
how about a structure like:

TemplateRenderer -> Renderer
AmpRenderer (naming TBD) -> Renderer

Both renderer import the friendlyframeutil to share code.

glevitzky replied:
My concern is that the util would just be the renderer itself, as there's little to nothing that would be left after extracting the common code. Can we circle back to this in a future PR?

lannka replied:
Having one renderer impl extend another renderer impl is the code pattern i want to avoid. Let's avoid at the beginning.

Given it's introduced in a previous PR, I'm OK to have this PR in first. But please do refactor in the next PR.

glevitzky replied:
Sure, I completely agree with you on avoiding a `TemplateRenderer -> FriendlyFrameRenderer -> Renderer` hierarchy -- that's why I changed TemplateRenderer to use an instance of FriendlyFrameRenderer rather than extend it.

But, yes, we can do a refactor/cleanup after this PR. Meanwhile, feel free to start reviewing this.

lannka commented on extensions/amp-a4a/0.1/template-config.js:
```diff
@@ -0,0 +1,30 @@
+export let NetworkRegistryDef;
+
+/** @const {!NetworkRegistryDef} */
+export const NetworkRegistry = {
```
I think the default config for ad networks can be a feature added later. As I'm yet not sure it will be helpful or not.

This `amp-ad-custom` extension is mainly for demo or trial purpose. Any real ad network should consider have its own extension `amp-ad-network-a`.

lannka commented on extensions/amp-a4a/0.1/amp-ad-template.js:
```diff
@@ -0,0 +1,87 @@
+    let url = this.baseRequestUrl_;
+
+    // We collect all fields in the dataset of the form
+    // 'data-request-var-<field_name>=<val>`, and append &<field_name>=<val> to
```
update doc here.

glevitzky replied:
Done.

lannka commented on extensions/amp-a4a/0.1/amp-ad-template.js:
```diff
@@ -0,0 +1,87 @@
+        }
+      }
+    });
+    this.getContext().adUrl = url;
```
is it needed right now?

I want to make sure we only add necessary stuff into the `context` object. And I wish we can later document clearly what the context object contains.

Can you check all the other context data, and remove the ones are not used.

glevitzky replied:
Yes, this is used by friendly-frame-renderer.js on line 97, though I'm not certain how necessary that is.

lannka commented on extensions/amp-ad-custom/0.1/amp-ad-custom.js:
```diff
@@ -0,0 +1,22 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {AmpAdTemplate} from '../../amp-a4a/0.1/amp-ad-template';
```
should we just move AmpAdTemplate here?

glevitzky replied:
Do you mean move the whole module into this file, or just move the module into this extension?

lannka replied:
whole module to this file. basically merge the 2 files.

glevitzky replied:
Done.

lannka commented on build-system/app.js:
```diff
@@ -933,6 +933,7 @@ app.get('/adzerk/*', (req, res) => {
   fs.readFileAsync(filePath).then(file => {
     res.setHeader('Content-Type', 'application/json');
     res.setHeader('AMP-template-amp-creative', 'amp-mustache');
```
I thought we already changed to
`AMP-Ad-Template-Extension: amp-mustache `

glevitzky replied:
We did, but I must've forgotten to change it here.

lannka replied:
and you can remove this line now

glevitzky replied:
Done.

lannka commented on extensions/amp-a4a/0.1/template-validator.js:
```diff
@@ -52,29 +52,31 @@ export class TemplateValidator extends Validator {
           /** @type {!./amp-ad-type-defs.ValidatorOutput} */ ({
             creativeData: {
-              creative: body,
+              crossDomainData: {
```
I'm trying to understand why this extra layer is needed?

glevitzky replied:
Can't recall if I had a reason or not. So far seems to work fine without it, so I've removed it.

lannka commented on extensions/amp-a4a/0.1/template-renderer.js:
```diff
@@ -15,34 +15,53 @@
  * Render AMP creative into FriendlyFrame via templatization.
  */
-export class TemplateRenderer extends FriendlyFrameRenderer {
+export class TemplateRenderer extends Renderer {
```
this is good. now we also need to change FriendlyFrameRender to a helper util class.

glevitzky replied:
Done.

lannka commented on extensions/amp-a4a/0.1/template-renderer.js:
```diff
@@ -15,34 +15,53 @@
+            return Promise.resolve();
+          }
+          const templateHelper = getAmpAdTemplateHelper(context.win);
+          const {iframe} = this.friendlyFrameRenderer_;
```
once you made `friendlyFrameRenderer` an util class, then its `render()` can return this `iframe`.

glevitzky replied:
Done.
